### PR TITLE
Help Center: use new Happychat availability endpoint

### PR DESCRIPTION
**Pull request by arcangelini**
## Proposed Changes

We recently have added an endpoint to Happychat that queries to see if there is availability. This PR updates the `useHappychatAvailable` to query this endpoint rather than opening a socket connection.

In order to make this work we had to add another endpoint relay that checks if the request is coming from staging or production in order to check availability against the correct environment. Check that here - D94755-code

## Testing Instructions

1. This needs to be tested in: Simple, Atomic, Editor, and WP-ADMIN.
2. Open the developer console and examine the Network tab to view the API calls being made.
3. Open Help Center and start a chat.
4. You should be able to close the chat and open it back up.
5. Refresh the screen to make sure the session persists.
6. Try and poke holes in it as best you can.

### CODE REVIEW

**Comment by arcangelini** [comment]
This is still pretty unrefined but I wanted to create the PR to open the dialogue. 💬

File: packages/happychat-connection/src/use-happychat-available.ts [hunk 1]
```diff
@@ -8,41 +8,19 @@ type HCAvailability = { available?: boolean; status?: string; env?: 'staging' |
 const key = Date.now();
 
 function getHCAvailabilityAndStatus( dataAuth: HappychatAuth ) {
+	const result: HCAvailability = {};
```
Comment (arcangelini) [inline packages/happychat-connection/src/use-happychat-available.ts]: One thing about modifying this function is that we will be fetching the authentication still with every request. I am not sure if we need to do that?

Maybe we can have two functions? `authenticateHappychat` and then `useHappychatAvailability`. That way we can set the stale time to something smaller like 60 seconds and check it more frequently without having to authenticate everytime.

Comment (alshakero) [reply packages/happychat-connection/src/use-happychat-available.ts:11]: > is that we will be fetching the authentication still with every request

Why is that?

File: packages/data-stores/src/support-queries/use-support-history.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+	if ( isLoading || ! tickets ) {
+		return false;
+	}
+
+	return tickets.length > 0;
+}
+
+export function useActiveSupportHistory( type: 'CHAT' | 'OTHER' ) {
+	return useQuery< SupportTicket >(
+		'help-support-history',
```
Comment (alshakero) [inline packages/data-stores/src/support-queries/use-support-history.ts]: You need to pass the type to the query key. Otherwise you'll get the wrong cached results.

```js
useActiveSupportHistory('CHAT'); // gives you CHAT and caches it
useActiveSupportHistory('OTHER'); // gives you CHAT, because react query got the same key
```

**Comment by arcangelini** [comment]
Pushed some changes. It took a while to sort out the useHasActiveSupport hook to run at the correct time when opening and closing the Help Center while chatting.

Next step is to tackle the `env` dilemma. We need to find a way to determine if the user is in prod or staging. This used to be passed down from the `/chat` session endpoint but now no longer exists. @alshakero suggested creating a new endpoint to determine this so I will work on that next.

**Comment by arcangelini** [comment]
Took a little bit of tinkering but I think we have a working model. I wired up the new endpoint to check the correct availability D94755-code we only need to sort out the typings. I struggled to type the `useHappychatAvailable` response [here](https://github.com/Automattic/wp-calypso/blob/dae000cceee70e3e79866b153b183200b598edc3/packages/happychat-connection/src/use-happychat-available.ts#L19-L20).

Next step is to update the availability url endpoint. We need to allow to pass a cache buster or build it in to the response. Right now we are not seeing the latest results I believe because of that.

We can start testing as of now I believe. We will need to be very thorough and test all the usual places Simple, Atomic, editor, calypso, wp-admin, etc.

**Comment by arcangelini** [comment]
Also, I wanted to highlight that I removed the `hidden` prop from the Help Center container. It was not being used for anything. I followed all the through the different areas we call the Help Center and found nothing.

**Review by alshakero — CHANGES_REQUESTED** [review]
We shouldn't leak the `proxied` concept to the client side. Let's keep `env` and not mention `proxied` at all.

I also have some feedback for the cache buster but need to validate tomorrow.

File: packages/data-stores/src/support-queries/use-support-history.ts (hunk 2 above)
Comment (arcangelini) [reply packages/data-stores/src/support-queries/use-support-history.ts]: I think this comment is outdated. I fixed this in the latest commit. Not sure where you saw this?

Comment (arcangelini) [reply packages/data-stores/src/support-queries/use-support-history.ts]: Can we call it `staging` and leave it a boolean? I switched to a boolean to make the logic simpler. That way proxied is not exposed.

**Comment by arcangelini** [comment]
> We shouldn't leak the `proxied` concept to the client side. Let's keep `env` and not mention `proxied` at all.
> 
> I also have some feedback for the cache buster but need to validate tomorrow.

Can we call it `staging` and leave it a boolean? I switched to a boolean to make the logic simpler. That way proxied is not exposed.

**Comment by alshakero** [comment]
Yes `staging` is great.

File: packages/data-stores/src/support-queries/use-support-history.ts (hunk 2 above)
Comment (arcangelini) [reply packages/data-stores/src/support-queries/use-support-history.ts]: This is what it currently is:

https://github.com/Automattic/wp-calypso/blob/719d86572789898249653cb42cac270e0c27d3f4/packages/data-stores/src/support-queries/use-support-history.ts#L14

**Comment by alshakero** [comment]
Important to run `yarn dev --sync` in `apps/happychat` and make sure it works before merging. When you all goes well, we have to deploy Happychat as well.

File: packages/happychat-connection/src/use-happychat-available.ts [hunk 3]
```diff
@@ -1,68 +1,25 @@
...
- */
-export function useHappychatAvailable( enabled = true, staleTime = 10 * 60 * 1000 ) {
-	const { data: dataAuth, isLoading: isLoadingAuth } = useHappychatAuth();
-
-	return useQuery(
-		'happychat-available' + key,
-		() => getHCAvailabilityAndStatus( dataAuth as HappychatAuth ),
+export function useHappychatAvailable( enabled = true ) {
+	return useQuery< HappychatAvailableResponse >(
+		`happychat-available-${ Math.floor( Date.now() / 1000 ) }`,
```
Comment (alshakero) [inline packages/happychat-connection/src/use-happychat-available.ts]: Although I have done this before, I don't recommend it. Because in Calypso, we persist things, so we'll persist these individually `happychat-available-4324324`. We should configure the query to refetch everytime (set all `refetchOn*` to `true`).

File: packages/help-center/src/components/help-center-container.tsx [hunk 4]
```diff
@@ -73,13 +73,13 @@ const HelpCenterContainer: React.FC< Container > = ( { handleClose, hidden } ) =
 	// https://github.com/react-grid-layout/react-draggable/blob/781ef77c86be9486400da9837f43b96186166e38/README.md
 	const nodeRef = useRef( null );
 
-	if ( ! show || hidden ) {
```
Comment (alshakero) [inline packages/help-center/src/components/help-center-container.tsx]: This is scope creep. We shouldn't have unrelated changes.

File: packages/help-center/src/components/help-center-notice.tsx [hunk 5]
```diff
@@ -113,7 +109,7 @@ export function HelpCenterActiveTicketNotice( {
 						/* translators: %s humanized date ex: 2 hours ago */
 						__( 'You submitted a request %s.', __i18n_text_domain__ ),
 						getRelativeTimeString( {
-							timestamp: tickets[ 0 ].timestamp * 1000,
+							timestamp: tickets.timestamp * 1000,
```
Comment (alshakero) [inline packages/help-center/src/components/help-center-notice.tsx]: If this is one ticket, we should rename the variable.

File: packages/help-center/src/components/help-center.tsx [hunk 6]
```diff
@@ -14,47 +13,37 @@ import { getSelectedSiteId } from 'calypso/state/ui/selectors';
...
 import { useHCWindowCommunicator } from '../happychat-window-communicator';
 import { useStillNeedHelpURL } from '../hooks/use-still-need-help-url';
-import { HELP_CENTER_STORE, USER_STORE, SITE_STORE } from '../stores';
+import { HELP_CENTER_STORE, SITE_STORE } from '../stores';
 import { Container } from '../types';
 import HelpCenterContainer from './help-center-container';
 
 import '../styles.scss';
 
-const HelpCenter: React.FC< Container > = ( { handleClose, hidden } ) => {
```
Comment (alshakero) [inline packages/help-center/src/components/help-center.tsx:23]: +1. Scope creep.

File: packages/help-center/src/hooks/use-should-render-chat-option.tsx [hunk 7]
```diff
@@ -6,16 +6,13 @@ type Result = {
 	state: 'AVAILABLE' | 'UNAVAILABLE' | 'CLOSED';
 	isLoading: boolean;
 	eligible: boolean;
-	env?: 'staging' | 'production';
+	env: boolean | undefined;
```
Comment (alshakero) [inline packages/help-center/src/hooks/use-should-render-chat-option.tsx]: since this is a boolean, `env` should be `isStaging` or so.

**Review by alshakero — CHANGES_REQUESTED** [review]
Tiny comments.

File: packages/happychat-connection/src/use-happychat-available.ts (hunk 3 above)
Comment (arcangelini) [reply packages/happychat-connection/src/use-happychat-available.ts]: I added a `refecthOnInterval` and set it to 30 seconds. By default I think all the `refetchOn*` are set to true.

I believe this will be really helpful to be able to check frequently as the user is filling out the form.

File: packages/happychat-connection/src/types.ts [hunk 8]
```diff
@@ -60,6 +60,8 @@ export interface HappychatAuth {
 	user: {
 		jwt: string;
 	} & HappychatUser;
+	is_existing_session: boolean;
+	availability_url: string;
```
Comment (alshakero) [inline packages/happychat-connection/src/types.ts]: are we using these?

File: packages/happychat-connection/src/use-happychat-available.ts [hunk 9]
```diff
@@ -1,68 +1,25 @@
...
-	const { data: dataAuth, isLoading: isLoadingAuth } = useHappychatAuth();
-
-	return useQuery(
-		'happychat-available' + key,
-		() => getHCAvailabilityAndStatus( dataAuth as HappychatAuth ),
+export function useHappychatAvailable( enabled = true ) {
+	return useQuery< HappychatAvailableResponse >(
+		'happychat-available',
+		() =>
+			apiFetch( {
```
Comment (alshakero) [inline packages/happychat-connection/src/use-happychat-available.ts:13]: can be just `fetch` without an import.

File: packages/happychat-connection/src/use-happychat-available.ts [hunk 10]
```diff
@@ -1,68 +1,25 @@
...
+				mode: 'cors',
+				method: 'GET',
+				credentials: 'omit',
+				url: 'https://public-api.wordpress.com/wpcom/v2/happychat/availability',
+			} ),
 		{
-			enabled: ! isLoadingAuth && !! dataAuth && enabled,
-			staleTime,
+			refetchInterval: 1000 * 30, // 30 seconds
```
Comment (alshakero) [inline packages/happychat-connection/src/use-happychat-available.ts:20]: Beautiful solution. But we should make sure we don't refresh every 30 seconds when the help center is closed.

**Review by alshakero — COMMENTED** [review]
Incredible work. Thank you so much. Code looks great. Will take it for a drive tomorrow.

File: packages/happychat-connection/src/types.ts (hunk 8 above)
Comment (arcangelini) [reply packages/happychat-connection/src/types.ts]: Nah, good catch this was leftovers from the original solution we had.

**Comment by arcangelini** [comment]
This is getting tabled until we can resolve the issue with Happychat sessions and how they are viewed as Open or Closed - pdDR7T-v7-p2

**Comment by arcangelini** [comment]
Closing this since we are moving towards ZD chat and this is now irrelevant
